app/gui/src/model/suggestion_database.rs

```diff
@@ -27,6 +28,50 @@ pub use example::Example;
 
 
 
+// ============================
+// === QualifiedNameToIdMap ===
+// ============================
+
+/// A map from fully qualified names (encoded as [`entry::QualifiedNameSegments`]) to
+/// [`entry::Id`]s. The methods of the type provide semantics of a map, while the internal
+/// representation is based on a [`HashMapTree`].
+///
+/// The internal representation conserves memory when storing many paths sharing a common prefix of
+/// segments.
+#[derive(Clone, Debug, Default)]
+struct QualifiedNameToIdMap {
+    tree: ensogl::data::HashMapTree<entry::NameSegment, Option<entry::Id>>,
+}
+
+impl QualifiedNameToIdMap {
+    fn warn_if_exists_and_set(&mut self, path: &entry::QualifiedNameSegments, id: entry::Id) {
+        let value = Some(id);
+        let old_value =
+            self.tree.replace_value_and_traverse_back_pruning_empty_leaf(&path.0, value);
+        if old_value.is_some() {
+            event!(WARN, "An existing suggestion entry id at {path:?} was overwritten with {id}.");
+        }
+    }
+
+    fn warn_if_absent_and_remove(&mut self, path: &entry::QualifiedNameSegments) {
+        let old_value = self.tree.replace_value_and_traverse_back_pruning_empty_leaf(&path.0, None);
+        if old_value.is_none() {
+            let msg = format!(
+                "Could not remove a suggestion entry id at {path:?} because it does not exist."
+            );
+            event!(WARN, "{msg}");
+        }
+    }
+
+    fn get<P, I>(&self, path: P) -> Option<entry::Id>
+    where
+        P: IntoIterator<Item = I>,
+        I: Into<entry::NameSegment>, {
+        self.tree.get(path).and_then(|v| *v)
+    }
```

Comment:
These functions (but `get`) can be added to `HashMapTree`. Then you don't need custom type here. But please, remember, that `HashMapTree` is like a `HashMap`. If you put `None` to `HashMap` then it exists in the hash map. Thus, this naming is incorrect.